tests/fuzz/runFuzzTest.sh

```diff
@@ -68,18 +70,68 @@ ACTUAL_TEST_FAILURE=0
 TEMP_CORPUS_DIR="$(mktemp -d)"
 cp -r ./corpus/${TEST_NAME}/. "${TEMP_CORPUS_DIR}"
 
-
 printf "Running %-s %-40s for %5d sec with %2d threads... " "${FIPS_TEST_MSG}" ${TEST_NAME} ${FUZZ_TIMEOUT_SEC} ${NUM_CPU_THREADS}
-./${TEST_NAME} ${LIBFUZZER_ARGS} ${TEMP_CORPUS_DIR} > ${TEST_NAME}_output.txt 2>&1 || ACTUAL_TEST_FAILURE=1
+
+# Setup and clean profile structure if FUZZ_COVERAGE is enabled, otherwise run as normal
+if [[ -v FUZZ_COVERAGE ]]; then
+    mkdir -p "./profiles/${TEST_NAME}"
+    rm -f ./profiles/${TEST_NAME}/*.profraw
+    LLVM_PROFILE_FILE="./profiles/${TEST_NAME}/${TEST_NAME}.%p.profraw" ./${TEST_NAME} ${LIBFUZZER_ARGS} ${TEMP_CORPUS_DIR} > ${TEST_NAME}_output.txt 2>&1 || ACTUAL_TEST_FAILURE=1
+else
+    ./${TEST_NAME} ${LIBFUZZER_ARGS} ${TEMP_CORPUS_DIR} > ${TEST_NAME}_output.txt 2>&1 || ACTUAL_TEST_FAILURE=1
+fi
+
 
 TEST_COUNT=`grep -o "stat::number_of_executed_units: [0-9]*" ${TEST_NAME}_output.txt | awk '{test_count += $2} END {print test_count}'`
 TESTS_PER_SEC=`echo $(($TEST_COUNT / $FUZZ_TIMEOUT_SEC))`
 FEATURE_COVERAGE=`grep -o "ft: [0-9]*" ${TEST_NAME}_output.txt | awk '{print $2}' | sort | tail -1`
+TARGET_FUNCS=''
+declare -i TARGET_TOTAL=0
+declare -i TARGET_COV=0
+
+# Outputs fuzz coverage results if the FUZZ_COVERAGE environment variable is set
+# Coverage is overlayed on source code in ${TEST_NAME}_cov.html, and coverage statistics are available in ${TEST_NAME}_cov.txt
+# If using LLVM version 9 or greater, coverage is output in LCOV format instead of HTML
+# All files are stored in the s2n coverage directory
+if [[ -v FUZZ_COVERAGE ]]; then
+    mkdir -p ${COVERAGE_DIR}/fuzz
+    llvm-profdata merge -sparse ./profiles/${TEST_NAME}/*.profraw -o ./profiles/${TEST_NAME}/${TEST_NAME}.profdata
+    llvm-cov report -instr-profile=./profiles/${TEST_NAME}/${TEST_NAME}.profdata ${S2N_ROOT}/lib/libs2n.so ${FUZZCOV_SOURCES} -show-functions > ${COVERAGE_DIR}/fuzz/${TEST_NAME}_cov.txt
+
+    # Use LCOV format instead of HTML if the LLVM version we're using supports it
+    if [[ $(grep -Eo "[0-9]*" <<< `llvm-cov --version` | head -1) > 8 ]]; then
+        llvm-cov show -instr-profile=./profiles/${TEST_NAME}/${TEST_NAME}.profdata ${S2N_ROOT}/lib/libs2n.so ${FUZZCOV_SOURCES} -use-color -format=lcov > ${COVERAGE_DIR}/fuzz/${TEST_NAME}_cov.info
+    else
+        llvm-cov show -instr-profile=./profiles/${TEST_NAME}/${TEST_NAME}.profdata ${S2N_ROOT}/lib/libs2n.so ${FUZZCOV_SOURCES} -use-color -format=html > ${COVERAGE_DIR}/fuzz/${TEST_NAME}_cov.html
+    fi
+
+    # Extract target functions from test source
+    TARGET_FUNCS=`sed -n "s|^/\* Target Functions: \([[:print:]]*\) \*/|\1|p" ${TEST_NAME}.c`
+
+    # Find line coverage statistics for target functions
+    if [[ ! -z TARGET_FUNCS ]];
```

Comment:
is this needed?

```
touch tmp
TARGET_FUNCS=`sed -n "s|^/\* Target Functions: \([[:print:]]*\) \*/|\1|p" tmp`
for TARGET in $TARGET_FUNCS; do echo "got \"$TARGET\""; done
```

doesn't output anything, so suggests this check is redundant.  but i'm using `zsh`, so maybe that's why...

Reply:
Good question! While the majority of shells currently behave properly with this, I can't guarantee that all shells will properly skip a for loop if the variable is null, or that the result of the sed command will always be assigned as a null value instead of a null string. This is just an extra check to prevent a possible error from a shell implementation not behaving as expected.